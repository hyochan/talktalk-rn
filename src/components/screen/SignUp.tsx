import React, { Component, useState, useEffect } from 'react';
import {
  StyleSheet,
  TouchableOpacity,
  Image,
  Text,
  View,
  ScrollView,
} from 'react-native';
import { NavigationScreenProp, NavigationStateRoute } from 'react-navigation';

import styled from 'styled-components/native';

import TextInput from '../shared/TextInput';
import Button from '../shared/Button';
import StatusBar from '../shared/StatusBar';

import { ratio, colors } from '../../utils/Styles';
import { getString } from '../../../STRINGS';

const StyledContainer = styled.View`
  flex:1;
  background-color: white;
  flex-direction: column;
  align-items: center;
`;

const StyledScrollView = styled.ScrollView`
  align-self: stretch;
`;

const StyledWrapper = styled.View`
  margin-top: 40;
  width: 78%;

  flex-direction: column;
  align-items: center;
`;

const StyledButtonWrapper = styled.View`
  margin-top: 24;
  width: 100%;
  align-items: flex-end;
`;

interface IProps {
  navigation: any;
}

interface IState {
  isRegistering: boolean;
  email: string;
  pw: string;
  displayName: string;
  statusMsg: string;
}

function Screen(props: IProps, state: IState) {
  const [isRegistering, setIsRegistering] = useState(false);
  const [email, setEmail] = useState('');
  const [pw, setPw] = useState('');
  const [displayName, setDisplayName] = useState('');
  const [statusMsg, setStatusMsg] = useState('');

  useEffect(() => {
    if (isRegistering) {
      setIsRegistering(false);
    }
  }, [isRegistering]);

  const onRegister = () => {
    setIsRegistering(true);
  };

  const onTextChanged = (type: string, text: string) => {
    switch (type) {
      case 'EMAIL':
        setEmail(text);
        break;
      case 'PW':
        setPw(text);
        break;
      case 'NAME':
        setDisplayName(text);
        break;
      case 'STATUS_MSG':
        setStatusMsg(text);
        break;
    }
  };

  return (
    <StyledContainer>
      <StatusBar isDarkContent={true}/>
      <StyledScrollView
        contentContainerStyle={{
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <StyledWrapper>
          <TextInput
            testID='email_input'
            style={{ marginTop: 8 }}
            txtLabel={getString('EMAIL')}
            txtHint={ getString('EMAIL') }
            txt={ email }
            onTextChanged={ (text) => onTextChanged('EMAIL', text)}
          />
          <TextInput
            testID='pw_input'
            style={{ marginTop: 24 }}
            txtLabel={ getString('PASSWORD') }
            txtHint={ getString('PASSWORD') }
            txt={ pw }
            onTextChanged={ (text) => onTextChanged('PW', text)}
            isPassword={ true }
          />
          <TextInput
            testID='name_input'
            style={{ marginTop: 24 }}
            txtLabel={getString('NAME')}
            txtHint={ getString('NAME') }
            txt={ displayName }
            onTextChanged={ (text) => onTextChanged('NAME', text)}
          />
          <TextInput
            testID='status_input'
            style={{ marginTop: 24 }}
            txtLabel={getString('STATUS_MSG')}
            txtHint={ getString('STATUS_MSG') }
            txt={ statusMsg }
            onTextChanged={ (text) => onTextChanged('STATUS_MSG', text)}
          />
          <StyledButtonWrapper>
            <Button
              testID='register'
              isLoading={isRegistering}
              onPress={() => onRegister()}
            >{getString('REGISTER')}</Button>
          </StyledButtonWrapper>
        </StyledWrapper>
      </StyledScrollView>
    </StyledContainer>
  );
}

export default Screen;
