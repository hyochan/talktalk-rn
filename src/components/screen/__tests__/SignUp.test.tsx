import 'react-native';
import * as React from 'react';
import SignUp from '../SignUp';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { render, fireEvent } from 'react-native-testing-library';

const props = {
  navigation: {
    navigate: jest.fn(),
  },
};
const component: React.ReactElement = <SignUp {...props}/>;

describe('[SignUp] rendering test', () => {
  it('renders as expected', () => {
    const json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
  });
});

describe('[SignUp] interaction', () => {
  let rendered: renderer.ReactTestRenderer;
  let testingLib: any;

  beforeAll(() => {
    rendered = renderer.create(component);
    testingLib = render(component);
  });

  it('should changeText when email changed', () => {
    const emailInput = testingLib.getByTestId('email_input');
    emailInput.props.onTextChanged('email test');
    expect(emailInput.props.txt).toEqual('email test');
  });

  it('should changeText when password changed', () => {
    const passwordInput = testingLib.getByTestId('pw_input');
    passwordInput.props.onTextChanged('pw test');
    expect(passwordInput.props.txt).toEqual('pw test');
  });

  it('should changeText when name changed', () => {
    const nameInput = testingLib.getByTestId('name_input');
    nameInput.props.onTextChanged('name test');
    expect(nameInput.props.txt).toEqual('name test');
  });

  it('should changeText when status changed', () => {
    const statusInput = testingLib.getByTestId('status_input');
    statusInput.props.onTextChanged('status test');
    expect(statusInput.props.txt).toEqual('status test');
  });

  it('should simulate when [onRegister] clicked', () => {
    const registerBtn = testingLib.getByTestId('register');
    fireEvent(registerBtn, 'press');
    expect(registerBtn.props.isLoading).toEqual(false);
    registerBtn.props.onPress();
    expect(registerBtn.props.isLoading).toEqual(true);
  });
});
