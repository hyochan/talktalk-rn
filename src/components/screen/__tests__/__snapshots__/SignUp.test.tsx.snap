// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SignUp] rendering test renders as expected 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
    style={
      Array [
        Object {
          "alignSelf": "stretch",
        },
      ]
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "marginTop": 40,
              "width": "78%",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-start",
                "alignSelf": "stretch",
                "flexDirection": "column",
                "justifyContent": "flex-start",
              },
              Object {
                "marginTop": 8,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "rgb(134,154,183)",
                "fontSize": 12,
                "marginBottom": 8,
              }
            }
          >
            Email
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            autoCorrect={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Email"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "alignSelf": "stretch",
                  "borderColor": "rgb(233,237,244)",
                  "borderWidth": 1,
                  "color": "rgb(65,77,107)",
                  "flexDirection": "row",
                  "fontSize": 16,
                  "justifyContent": "flex-start",
                  "paddingHorizontal": 20,
                  "paddingVertical": 22,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-start",
                "alignSelf": "stretch",
                "flexDirection": "column",
                "justifyContent": "flex-start",
              },
              Object {
                "marginTop": 24,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "rgb(134,154,183)",
                "fontSize": 12,
                "marginBottom": 8,
              }
            }
          >
            Password
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            autoCorrect={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Password"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "alignSelf": "stretch",
                  "borderColor": "rgb(233,237,244)",
                  "borderWidth": 1,
                  "color": "rgb(65,77,107)",
                  "flexDirection": "row",
                  "fontSize": 16,
                  "justifyContent": "flex-start",
                  "paddingHorizontal": 20,
                  "paddingVertical": 22,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-start",
                "alignSelf": "stretch",
                "flexDirection": "column",
                "justifyContent": "flex-start",
              },
              Object {
                "marginTop": 24,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "rgb(134,154,183)",
                "fontSize": 12,
                "marginBottom": 8,
              }
            }
          >
            name
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            autoCorrect={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="name"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "alignSelf": "stretch",
                  "borderColor": "rgb(233,237,244)",
                  "borderWidth": 1,
                  "color": "rgb(65,77,107)",
                  "flexDirection": "row",
                  "fontSize": 16,
                  "justifyContent": "flex-start",
                  "paddingHorizontal": 20,
                  "paddingVertical": 22,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-start",
                "alignSelf": "stretch",
                "flexDirection": "column",
                "justifyContent": "flex-start",
              },
              Object {
                "marginTop": 24,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "rgb(134,154,183)",
                "fontSize": 12,
                "marginBottom": 8,
              }
            }
          >
            Status message
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            autoCorrect={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Status message"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "alignSelf": "stretch",
                  "borderColor": "rgb(233,237,244)",
                  "borderWidth": 1,
                  "color": "rgb(65,77,107)",
                  "flexDirection": "row",
                  "fontSize": 16,
                  "justifyContent": "flex-start",
                  "paddingHorizontal": 20,
                  "paddingVertical": 22,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "marginTop": 24,
                "width": "100%",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                },
              ]
            }
          >
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="press_id"
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "rgb(58,139,255)",
                      "borderColor": "rgb(58,139,255)",
                      "borderRadius": 4,
                      "borderWidth": 1,
                      "justifyContent": "center",
                      "shadowColor": "rgb(58,139,255)",
                      "shadowOpacity": 0.3,
                      "shadowRadius": 4,
                    },
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "height": 60,
                      "justifyContent": "center",
                      "width": 136,
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "white",
                          "fontSize": 14,
                          "fontWeight": "bold",
                        },
                      ]
                    }
                  >
                    Register
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
